# Importing libraries

In [7]:
#Importing Libraries 
import pandas as pd
import numpy as np
import os 

path = r'C:\Users\ANITA BOADU\Achievement 4'

# Import the dataframe ords_prods_merge as 
ords_prods_merge = pd.read_pickle(os.path.join(path,'Data','Prepared Data','ords_prods_merge.pkl'))



In [8]:
ords_prods_merge.shape

(30356957, 16)

In [9]:
ords_prods_merge.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           1   2398795        1    prior             2          3   
1           1   2398795        1    prior             2          3   
2           1   2398795        1    prior             2          3   
3           1   2398795        1    prior             2          3   
4           1   2398795        1    prior             2          3   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  7                    15.0         196                  1   
1                  7                    15.0       10258                  2   
2                  7                    15.0       12427                  3   
3                  7                    15.0       13176                  4   
4                  7                    15.0       26088                  5   

   reordered _merge                product_name  aisle_id  department_id  \
0          1   both                        Soda      77.0            7.0   
1          0   both                  Pistachios     117.0           19.0   
2          1   both         Original Beef Jerky      23.0           19.0   
3          0   both      Bag of Organic Bananas      24.0            4.0   
4          1   both  Aged White Cheddar Popcorn      23.0           19.0   

   prices  
0     9.0  
1     3.0  
2     4.4  
3    10.3  
4     4.7

In [10]:
ords_prods_merge.groupby('product_name')

###########Creates a grouped object where all rows with the same product_name
are grouped together.By itself, this command does not produce a visible result.
It prepares the data so you can apply operations like counting, summing, or aggregating.

In [11]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              16.559358
2.0              18.413176
3.0              18.279600
4.0              18.915890
5.0              16.497751
6.0              17.609390
7.0              18.303975
8.0              16.383301
9.0              17.022963
10.0             21.227447
11.0             17.311768
12.0             16.953613
13.0             17.727470
14.0             17.898097
15.0             17.290123
16.0             18.757796
17.0             16.808683
18.0             20.336389
19.0             18.258620
20.0             17.578959
21.0             23.644477

#######Groups rows based on department_id and applies the mean function to the order_number column
Returns a table showing the average order number for each department

In [12]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1.0     16.559358
2.0     18.413176
3.0     18.279600
4.0     18.915890
5.0     16.497751
6.0     17.609390
7.0     18.303975
8.0     16.383301
9.0     17.022963
10.0    21.227447
11.0    17.311768
12.0    16.953613
13.0    17.727470
14.0    17.898097
15.0    17.290123
16.0    18.757796
17.0    16.808683
18.0    20.336389
19.0    18.258620
20.0    17.578959
21.0    23.644477
Name: order_number, dtype: float64

In [20]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1.0              16.559358   2  99
2.0              18.413176   2  99
3.0              18.279600   2  99
4.0              18.915890   2  99
5.0              16.497751   2  99
6.0              17.609390   2  99
7.0              18.303975   2  99
8.0              16.383301   2  99
9.0              17.022963   2  99
10.0             21.227447   2  99
11.0             17.311768   2  99
12.0             16.953613   2  99
13.0             17.727470   2  99
14.0             17.898097   2  99
15.0             17.290123   2  99
16.0             18.757796   2  99
17.0             16.808683   2  99
18.0             20.336389   2  99
19.0             18.258620   2  99
20.0             17.578959   2  99
21.0             23.644477   2  99

########Groups all rows by department_id and For each department, it calculates:Mean (average order number), Minimum order number and the Maximum order number. The result is a table showing how order numbers vary across different departments.

In [13]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('max')

In [13]:
ords_prods_merge.head(15)

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0            1   2398795        1    prior             2          3   
1            1   2398795        1    prior             2          3   
2            1   2398795        1    prior             2          3   
3            1   2398795        1    prior             2          3   
4            1   2398795        1    prior             2          3   
5            1   2398795        1    prior             2          3   
6            2    473747        1    prior             3          3   
7            2    473747        1    prior             3          3   
8            2    473747        1    prior             3          3   
9            2    473747        1    prior             3          3   
10           2    473747        1    prior             3          3   
11           3   2254736        1    prior             4          4   
12           3   2254736        1    prior             4          4   
13           3   2254736        1    prior             4          4   
14           3   2254736        1    prior             4          4   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   7                    15.0         196                  1   
1                   7                    15.0       10258                  2   
2                   7                    15.0       12427                  3   
3                   7                    15.0       13176                  4   
4                   7                    15.0       26088                  5   
5                   7                    15.0       13032                  6   
6                  12                    21.0         196                  1   
7                  12                    21.0       12427                  2   
8                  12                    21.0       10258                  3   
9                  12                    21.0       25133                  4   
10                 12                    21.0       30450                  5   
11                  7                    29.0         196                  1   
12                  7                    29.0       12427                  2   
13                  7                    29.0       10258                  3   
14                  7                    29.0       25133                  4   

    reordered _merge                product_name  aisle_id  department_id  \
0           1   both                        Soda      77.0            7.0   
1           0   both                  Pistachios     117.0           19.0   
2           1   both         Original Beef Jerky      23.0           19.0   
3           0   both      Bag of Organic Bananas      24.0            4.0   
4           1   both  Aged White Cheddar Popcorn      23.0           19.0   
5           0   both       Cinnamon Toast Crunch     121.0           14.0   
6           1   both                        Soda      77.0            7.0   
7           1   both         Original Beef Jerky      23.0           19.0   
8           1   both                  Pistachios     117.0           19.0   
9           0   both       Organic String Cheese      21.0           16.0   
10          0   both        Creamy Almond Butter      88.0           13.0   
11          1   both                        Soda      77.0            7.0   
12          1   both         Original Beef Jerky      23.0           19.0   
13          1   both                  Pistachios     117.0           19.0   
14          1   both       Organic String Cheese      21.0           16.0   

    prices  max_order  
0      9.0         10  
1      3.0         10  
2      4.4         10  
3     10.3         10  
4      4.7         10  
5      4.0         10  
6      9.0         10  
7      4.4         10  
8      3.0         10  
9      8.6         10  
10     3.2         10  
11     9.0         10  
12     4.4         10  
13     3.0       

In [14]:
ords_prods_merge.head(100)

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0            1   2398795        1    prior             2          3   
1            1   2398795        1    prior             2          3   
2            1   2398795        1    prior             2          3   
3            1   2398795        1    prior             2          3   
4            1   2398795        1    prior             2          3   
..         ...       ...      ...      ...           ...        ...   
95          16   1199898        2    prior             6          2   
96          16   1199898        2    prior             6          2   
97          16   1199898        2    prior             6          2   
98          16   1199898        2    prior             6          2   
99          16   1199898        2    prior             6          2   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   7                    15.0         196                  1   
1                   7                    15.0       10258                  2   
2                   7                    15.0       12427                  3   
3                   7                    15.0       13176                  4   
4                   7                    15.0       26088                  5   
..                ...                     ...         ...                ...   
95                  9                    13.0       32052                  5   
96                  9                    13.0       27344                  6   
97                  9                    13.0       24852                  7   
98                  9                    13.0       45066                  8   
99                  9                    13.0       27966                  9   

    reordered _merge                       product_name  aisle_id  \
0           1   both                               Soda      77.0   
1           0   both                         Pistachios     117.0   
2           1   both                Original Beef Jerky      23.0   
3           0   both             Bag of Organic Bananas      24.0   
4           1   both         Aged White Cheddar Popcorn      23.0   
..        ...    ...                                ...       ...   
95          0   both  Organic Brown Rice Cake Salt-Free      78.0   
96          1   both               Uncured Genoa Salami      96.0   
97          1   both                             Banana      24.0   
98          1   both                   Honeycrisp Apple      24.0   
99          1   both                Organic Raspberries     123.0   

    department_id  prices  max_order  
0             7.0     9.0         10  
1            19.0     3.0         10  
2            19.0     4.4         10  
3             4.0    10.3         10  
4            19.0     4.7         10  
..            ...     ...        ...  
95           19.0     4.3         14  
96           20.0    14.8         14  
97            4.0    12.3         14  
98            4.0     2.5         14  
99            4.0     4.4         14  

[100 rows x 17 columns]

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [24]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [17]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
NaN               15095623
Loyal customer    10105134
New customer       5156200
Name: count, dtype: int64

########The numbers represent how many orders or customers fall into each loyalty category.
Most customers are regular shoppers, but a large portion are loyal, indicating a healthy returning customer base.
The new customers show ongoing growth potential.

########In the subset, the **'Regular customer'** category is properly labeled, but in the entire DataFrame, it appears as **NaN**.  
   - This likely indicates that the `loyalty_flag` column in the full DataFrame has missing values for those rows that correspond to regular customers in the subset.  
   - The subset may have been filtered to exclude missing values, which is why the label shows correctly there.

2. The counts for **Loyal** and **New customers** remain the same in both the subset and entire DataFrame.  
   - This suggests that these categories were fully populated and consistent across the full dataset.

3. Overall, the main difference is **missing labels (NaN) in the full DataFrame** for Regular customers, which does not affect the actual numbers but affects readability and labeling.


In [18]:
ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                             
Loyal customer  10095381.0  10.402162  329.036155  1.0  4.2  7.4  11.2   
New customer     5151691.0  13.396333  604.763039  1.0  4.2  7.4  11.3   

                    max  
loyalty_flag             
Loyal customer  99999.0  
New customer    99999.0

#######Mean prices appear higher for new customers compared to loyal customers 
Median and quartiles are very similar across groups (median = 7.4), indicating that the typical purchase price is similar.
Maximum values (99,999) are likely outliers and are inflating the mean.Typical spending is comparable between loyal and new customers.

In [19]:
user_avg_price = ords_prods_merge.groupby('user_id')['prices'].mean().reset_index()

#######We first group the data by user_id and calculate the mean price of products purchased by each user.

In [20]:
def spending_flag(avg_price):
    if avg_price < 10:
        return "Low spender"
    else:
        return "High spender"

Users with average price < 10 → "Low spender"

Users with average price ≥ 10 → "High spender"

In [21]:
user_avg_price.rename(columns={'prices': 'avg_price'}, inplace=True)

#######Rename the prices column to avg_price for easier understanding.

In [22]:
user_avg_price['spending_flag'] = user_avg_price['avg_price'].apply(spending_flag)

#######Create a new column spending_flag by applying the function to the avg_price column.

In [23]:
ords_prods_merge = ords_prods_merge.merge(user_avg_price[['user_id', 'spending_flag']], on='user_id', how='left'
)

########Every row in ords_prods_merge now has a spending_flag column

In [24]:
print(ords_prods_merge['spending_flag'].value_counts())

spending_flag
Low spender     29758265
High spender      598692
Name: count, dtype: int64


########This shows how many users are Low spenders vs High spenders.The majority of users (≈ 29.8 million orders) fall into the Low spender category.Only a small fraction of users (≈ 0.6 million orders) are High spenders.
This indicates that most customers purchase lower-priced items, while a small group contributes to higher-value sales.

In [25]:
user_order_median = ords_prods_merge.groupby('user_id')['days_since_prior_order'].median().reset_index()


#######We first calculate the median number of days between orders for each user. This gives a better picture of their typical ordering behavior than the mean, because it is robust to outliers.

In [26]:
def order_frequency_flag(median_days):
    if median_days > 20:
        return "Non-frequent customer"
    elif median_days > 10 and median_days <= 20:
        return "Regular customer"
    else:
        return "Frequent customer"

The flag is defined using the following criteria: median_days > 20 → "Non-frequent customer" median_days > 10 and <= 20 → "Regular customer" median_days <= 10 → "Frequent customer"

In [27]:
user_order_median['order_frequency_flag'] = user_order_median['days_since_prior_order'].apply(order_frequency_flag)


########Every row in ords_prods_merge now contains the order_frequency_flag.

In [28]:
ords_prods_merge = ords_prods_merge.merge(user_order_median[['user_id', 'order_frequency_flag']],on='user_id', how='left'
)

In [29]:
print(ords_prods_merge['order_frequency_flag'].value_counts())

order_frequency_flag
Frequent customer        20692982
Regular customer          6602393
Non-frequent customer     3061582
Name: count, dtype: int64


########This shows how many users are Frequent, Regular, or Non-frequent customers.

In [30]:
ords_prods_merge.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           1   2398795        1    prior             2          3   
1           1   2398795        1    prior             2          3   
2           1   2398795        1    prior             2          3   
3           1   2398795        1    prior             2          3   
4           1   2398795        1    prior             2          3   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  7                    15.0         196                  1   
1                  7                    15.0       10258                  2   
2                  7                    15.0       12427                  3   
3                  7                    15.0       13176                  4   
4                  7                    15.0       26088                  5   

   reordered _merge                product_name  aisle_id  department_id  \
0          1   both                        Soda      77.0            7.0   
1          0   both                  Pistachios     117.0           19.0   
2          1   both         Original Beef Jerky      23.0           19.0   
3          0   both      Bag of Organic Bananas      24.0            4.0   
4          1   both  Aged White Cheddar Popcorn      23.0           19.0   

   prices  max_order  loyalty_flag spending_flag   order_frequency_flag  
0     9.0         10  New customer   Low spender  Non-frequent customer  
1     3.0         10  New customer   Low spender  Non-frequent customer  
2     4.4         10  New customer   Low spender  Non-frequent customer  
3    10.3         10  New customer   Low spender  Non-frequent customer  
4     4.7         10  New customer   Low spender  Non-frequent customer

In [31]:
ords_prods_merge.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.035696e+07  3.035696e+07  3.035696e+07  3.035696e+07  3.035696e+07   
mean   1.709803e+06  1.710736e+06  1.029339e+05  1.824695e+01  2.740779e+00   
std    9.865039e+05  9.872460e+05  5.946892e+04  1.759162e+01  2.087926e+00   
min    1.000000e+00  2.000000e+00  1.000000e+00  2.000000e+00  0.000000e+00   
25%    8.554820e+05  8.561350e+05  5.140900e+04  6.000000e+00  1.000000e+00   
50%    1.708865e+06  1.710892e+06  1.025960e+05  1.200000e+01  3.000000e+00   
75%    2.565805e+06  2.565520e+06  1.543850e+05  2.500000e+01  5.000000e+00   
max    3.421081e+06  3.421083e+06  2.062090e+05  9.900000e+01  6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.035696e+07            3.035696e+07  3.035696e+07   
mean        1.341124e+01            1.110414e+01  2.557424e+04   
std         4.246811e+00            8.778941e+00  1.409443e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.351900e+04   
50%         1.300000e+01            8.000000e+00  2.523700e+04   
75%         1.600000e+01            1.500000e+01  3.793300e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.035696e+07  3.035696e+07  3.032876e+07   3.032876e+07   
mean        8.354155e+00  6.300634e-01  7.119183e+01   9.922036e+00   
std         7.133524e+00  4.827872e-01  3.821413e+01   6.281967e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices     max_order  
count  3.032876e+07  3.035696e+07  
mean   1.197725e+01  3.420227e+01  
std    4.963730e+02  2.521065e+01  
min    1.000000e+00  3.000000e+00  
25%    4.200000e+00  1.400000e+01  
50%    7.400000e+00  2.800000e+01  
75%    1.130000e+01  4.800000e+01  
max    9.999900e+04  9.900000e+01

In [32]:
ords_prods_merge.columns

Index(['Unnamed: 0', 'order_id', 'user_id', 'eval_set', 'order_number',
       'order_dow', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', '_merge',
       'product_name', 'aisle_id', 'department_id', 'prices', 'max_order',
       'loyalty_flag', 'spending_flag', 'order_frequency_flag'],
      dtype='object')

In [33]:
ords_prods_merge.shape

(30356957, 20)

In [34]:
ords_prods_merge.to_pickle(
    os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_4.8.pkl')
)

In [35]:
test = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_4.8.pkl'))

In [36]:
print(test.columns)

Index(['Unnamed: 0', 'order_id', 'user_id', 'eval_set', 'order_number',
       'order_dow', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', '_merge',
       'product_name', 'aisle_id', 'department_id', 'prices', 'max_order',
       'loyalty_flag', 'spending_flag', 'order_frequency_flag'],
      dtype='object')


In [37]:
print(test.shape)

(30356957, 20)
